In [18]:
import pandas as pd
df=pd.read_csv("heart.csv")
df.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

In [19]:
df.describe()

Age   RestingBP  Cholesterol   FastingBS       MaxHR  \
count  918.000000  918.000000   918.000000  918.000000  918.000000   
mean    53.510893  132.396514   198.799564    0.233115  136.809368   
std      9.432617   18.514154   109.384145    0.423046   25.460334   
min     28.000000    0.000000     0.000000    0.000000   60.000000   
25%     47.000000  120.000000   173.250000    0.000000  120.000000   
50%     54.000000  130.000000   223.000000    0.000000  138.000000   
75%     60.000000  140.000000   267.000000    0.000000  156.000000   
max     77.000000  200.000000   603.000000    1.000000  202.000000   

          Oldpeak  HeartDisease  
count  918.000000    918.000000  
mean     0.887364      0.553377  
std      1.066570      0.497414  
min     -2.600000      0.000000  
25%      0.000000      0.000000  
50%      0.600000      1.000000  
75%      1.500000      1.000000  
max      6.200000      1.000000

In [20]:
df.shape

(918, 12)

In [21]:
df['zscore_Cholesterol']=(df.Cholesterol-(df.Cholesterol.mean()))/df.Cholesterol.std()
                           

In [22]:
df['zscore_Cholesterol']

0      0.824621
1     -0.171867
2      0.769768
3      0.138964
4     -0.034736
         ...   
913    0.596068
914   -0.053020
915   -0.619830
916    0.340090
917   -0.217578
Name: zscore_Cholesterol, Length: 918, dtype: float64

## FInd the cholestrol outliers having zscore >3 and <-3 

In [32]:
import warnings
warnings.filterwarnings('ignore')

In [33]:

zscore=(df[(df['zscore_Cholesterol']>3)|(df['zscore_Cholesterol']<-3)])
zscore

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
76    32   M           ASY        118          529          0     Normal   
149   54   M           ASY        130          603          1     Normal   
616   67   F           NAP        115          564          0        LVH   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  zscore_Cholesterol  
76     130              N      0.0     Flat             1            3.018723  
149    125              Y      1.0     Flat             1            3.695238  
616    160              N      1.6     Flat             0            3.338696

In [34]:
df1=df[(df['zscore_Cholesterol']<3)&(df['zscore_Cholesterol']>-3)]
df1.shape

(915, 13)

In [35]:
df.shape

(918, 13)

In [36]:
df1['Zscore_MaxHR']=(df1.MaxHR-(df1.MaxHR.mean()))/df1.MaxHR.std()
df1


Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  zscore_Cholesterol  \
0      172              N      0.0       Up             0            0.824621   
1      156              N      1.0     Flat             1           -0.171867   
2       98              N      0.0       Up             0            0.769768   
3      108              Y      1.5     Flat             1            0.138964   
4      122              N      0.0       Up             0           -0.034736   
..     ...            ...      ...      ...           ...                 ...   
913    132              N      1.2     Flat             1            0.596068   
914    141              N      3.4     Flat             1           -0.053020   
915    115              Y      1.2     Flat             1           -0.619830   
916    174              N      0.0     Flat             1            0.340090   
917    173              N      0.0       Up             0           -0.217578   

     Zscore_MaxHR  
0        1.380949  
1        0.753167  
2       -1.522543  
3       -1.130179  
4       -0.580870  
..            ...  
913     -0.188506  
914      0.164621  
915     -0.855524  
916      1.459422  
917      1.420185  

[915 rows x 14 columns]

In [37]:
zscore=(df1[(df1['Zscore_MaxHR']>3)|(df1['Zscore_MaxHR']<-3)])
zscore

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
390   51   M           ASY        140            0          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  zscore_Cholesterol  \
390     60              N      0.0     Flat             1           -1.817444   

     Zscore_MaxHR  
390     -3.013525

## Remove outliers of MaxHR

In [39]:
df2=(df1[(df1['Zscore_MaxHR']<3)&(df1['Zscore_MaxHR']>-3)])
df2.shape

(914, 14)

## FInd Zscore for FastingBS feature

In [40]:
df2['zscore_FastingBS']=(df2.FastingBS-(df2.FastingBS.mean()))/df2.FastingBS.std()
df2['zscore_FastingBS']

0     -0.550926
1     -0.550926
2     -0.550926
3     -0.550926
4     -0.550926
         ...   
913   -0.550926
914    1.813141
915   -0.550926
916   -0.550926
917   -0.550926
Name: zscore_FastingBS, Length: 914, dtype: float64

## Remove the tuples having zscore_FastingBS value >3 and <-3

In [41]:
df3=(df2[(df2['zscore_FastingBS']<3)&(df2['zscore_FastingBS']>-3)])
df3.shape

(914, 15)

## Find Zscore for Oldpeak feature

In [42]:
df3['zscore_Oldpeak']=(df3.Oldpeak-(df3.Oldpeak.mean()))/df3.Oldpeak.std()

In [44]:
df4=(df3[(df3['zscore_Oldpeak']<3)&(df3['zscore_Oldpeak']>-3)])
df4.shape

(907, 16)

## Find Zscore for RestingBP feature and remove outliers

In [45]:
df4['zscore_RestingBP']=(df4.RestingBP-(df4.RestingBP.mean()))/df4.RestingBP.std()

In [47]:
df5=(df4[(df4['zscore_RestingBP']<3)&(df4['zscore_RestingBP']>-3)])
df5.shape

(899, 17)

In [49]:
df5.sample(5)

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
444   56   M           ASY        120          100          0     Normal   
762   40   M           ASY        110          167          0        LVH   
208   28   M           ATA        130          132          0        LVH   
124   58   M           ATA        130          230          0     Normal   
622   59   M           ASY        110          239          0        LVH   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  zscore_Cholesterol  \
444    120              Y      1.5     Flat             1           -0.903235   
762    114              Y      2.0     Flat             1           -0.290715   
208    185              N      0.0       Up             0           -0.610688   
124    150              N      0.0       Up             0            0.285237   
622    142              Y      1.2     Flat             1            0.367516   

     Zscore_MaxHR  zscore_FastingBS  zscore_Oldpeak  zscore_RestingBP  
444     -0.659343         -0.550926        0.572748         -0.669762  
762     -0.894761         -0.550926        1.040988         -1.210941  
208      1.891022         -0.550926       -0.831971         -0.128582  
124      0.517749         -0.550926       -0.831971         -0.128582  
622      0.203858         -0.550926        0.291804         -1.210941

In [50]:
df5.ChestPainType.unique()

array(['ATA', 'NAP', 'ASY', 'TA'], dtype=object)

In [51]:
df5.RestingECG.unique()

array(['Normal', 'ST', 'LVH'], dtype=object)

In [52]:
df5.ExerciseAngina.unique()

array(['N', 'Y'], dtype=object)

In [53]:
df5.ST_Slope.unique()

array(['Up', 'Flat', 'Down'], dtype=object)

### Handle text columns using label encoding and onehot encoding

In [55]:
df6=df5.copy()
df6.ExerciseAngina.replace(
    { 'N':0,
     'Y':1},inplace=True)
df6.RestingECG.replace(
    {'Normal':1,
     'ST':2,
     'LVH':3},inplace=True)
df6.ST_Slope.replace(
    {'Up':3, 
     'Flat':2,
     'Down':1},inplace=True)
df6.head(5)
    

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS  RestingECG  \
0   40   M           ATA        140          289          0           1   
1   49   F           NAP        160          180          0           1   
2   37   M           ATA        130          283          0           2   
3   48   F           ASY        138          214          0           1   
4   54   M           NAP        150          195          0           1   

   MaxHR  ExerciseAngina  Oldpeak  ST_Slope  HeartDisease  zscore_Cholesterol  \
0    172               0      0.0         3             0            0.824621   
1    156               0      1.0         2             1           -0.171867   
2     98               0      0.0         3             0            0.769768   
3    108               1      1.5         2             1            0.138964   
4    122               0      0.0         3             0           -0.034736   

   Zscore_MaxHR  zscore_FastingBS  zscore_Oldpeak  zscore_RestingBP  
0      1.380949         -0.550926       -0.831971          0.412597  
1      0.753167         -0.550926        0.104509          1.494956  
2     -1.522543         -0.550926       -0.831971         -0.128582  
3     -1.130179         -0.550926        0.572748          0.304361  
4     -0.580870         -0.550926       -0.831971          0.953777

In [56]:
df6.shape

(899, 17)

In [57]:
df7 =pd.get_dummies(df6,drop_first=True)
df7.shape

(899, 19)

In [58]:
df7.head(5)

Age  RestingBP  Cholesterol  FastingBS  RestingECG  MaxHR  ExerciseAngina  \
0   40        140          289          0           1    172               0   
1   49        160          180          0           1    156               0   
2   37        130          283          0           2     98               0   
3   48        138          214          0           1    108               1   
4   54        150          195          0           1    122               0   

   Oldpeak  ST_Slope  HeartDisease  zscore_Cholesterol  Zscore_MaxHR  \
0      0.0         3             0            0.824621      1.380949   
1      1.0         2             1           -0.171867      0.753167   
2      0.0         3             0            0.769768     -1.522543   
3      1.5         2             1            0.138964     -1.130179   
4      0.0         3             0           -0.034736     -0.580870   

   zscore_FastingBS  zscore_Oldpeak  zscore_RestingBP  Sex_M  \
0         -0.550926       -0.831971          0.412597      1   
1         -0.550926        0.104509          1.494956      0   
2         -0.550926       -0.831971         -0.128582      1   
3         -0.550926        0.572748          0.304361      0   
4         -0.550926       -0.831971          0.953777      1   

   ChestPainType_ATA  ChestPainType_NAP  ChestPainType_TA  
0                  1                  0                 0  
1                  0                  1                 0  
2                  1                  0                 0  
3                  0                  0                 0  
4                  0                  1                 0

In [61]:
df7=df7.drop(['zscore_Cholesterol','Zscore_MaxHR','zscore_FastingBS','zscore_Oldpeak','zscore_RestingBP'],axis='columns')


KeyError: "['zscore_Cholesterol', 'Zscore_MaxHR', 'zscore_FastingBS', 'zscore_Oldpeak', 'zscore_RestingBP'] not found in axis"

In [62]:
df7.shape

(899, 14)

In [63]:
x=df7.drop(['HeartDisease'],axis='columns')
y=df7.HeartDisease
x.shape

(899, 13)

In [66]:
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
x_scaled=scaler.fit_transform(x)
x_scaled.shape

(899, 13)

In [67]:
x_scaled

array([[-1.42815446,  0.46590022,  0.84963584, ...,  2.06332497,
        -0.5349047 , -0.22955001],
       [-0.47585532,  1.63471366, -0.16812204, ..., -0.48465463,
         1.86949191, -0.22955001],
       [-1.7455875 , -0.1185065 ,  0.79361247, ...,  2.06332497,
        -0.5349047 , -0.22955001],
       ...,
       [ 0.3706328 , -0.1185065 , -0.62564622, ..., -0.48465463,
        -0.5349047 , -0.22955001],
       [ 0.3706328 , -0.1185065 ,  0.35476274, ...,  2.06332497,
        -0.5349047 , -0.22955001],
       [-1.63977649,  0.34901888, -0.21480818, ..., -0.48465463,
         1.86949191, -0.22955001]])

In [68]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x_scaled,y,test_size=0.2,random_state=20)

In [69]:
x_train.shape

(719, 13)

In [70]:
x_test.shape

(180, 13)

## Train a model using standalone support vector machine and then using bagging

In [71]:
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_score
scores=cross_val_score(SVC(),x_scaled,y,cv=5)
scores.mean()

0.8152948479205462

## use bagging with SVM

In [72]:
from sklearn.ensemble import BaggingClassifier
bag_model=BaggingClassifier(base_estimator=SVC(),n_estimators=100,max_samples=0.8,random_state=0)
scores=cross_val_score(bag_model,x_scaled,y,cv=5)
scores.mean()

0.8164059590316572

#### As you can see above, using bagging in case of SVM doesn't make much difference in terms of model accuracy. Bagging is effective when we have high variance and instable model such as decision tree. Let's explore how bagging changes the performance for a decision tree classifier.

## Train a model using decision tree and then bagging


In [73]:
from sklearn.tree import DecisionTreeClassifier
scores=cross_val_score(DecisionTreeClassifier(),x_scaled,y,cv=5)
scores.mean()

0.7084357541899442

In [74]:
bag_model=BaggingClassifier(base_estimator=DecisionTreeClassifier(),n_estimators=100,max_samples=0.8,random_state=0)
scores=cross_val_score(bag_model,x_scaled,y,cv=5)
scores.mean()

0.8019056486654252

### The decision tree model's accuracy is improved from 70.84% to 80.19% when including the bagging

### Train a model using Random Forest which itself uses bagging underneath

In [75]:
from sklearn.ensemble import RandomForestClassifier
scores=cross_val_score(RandomForestClassifier(),x_scaled,y,cv=5)
scores.mean()

0.8252948479205463

#### Random forest gave even a better performance with 81.7% as score. Underneath it used bagging where it sampled not only data rows but also the columns (or features)